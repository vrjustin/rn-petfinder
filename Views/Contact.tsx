import React from 'react';
import {View, Text, StyleSheet, Linking, TouchableOpacity} from 'react-native';
import en from '../strings/en.json';

const Contact = ({route}) => {
  const {selectedAnimal} = route.params;
  const renderContactInfo = () => {
    const {contact} = selectedAnimal;
    if (!contact) {
      return null;
    }

    const {address, email, phone} = contact;

    // Check if address details are sufficient and not a PO Box
    const hasValidAddress =
      address &&
      address.address1 &&
      address.city &&
      address.state &&
      address.postcode &&
      !/^(P(\.|ost)?\s?O(\.|ffice)?\s?Box|POB?)\s*\d+/i.test(address.address1);

    return (
      <View style={styles.contactContainer}>
        {hasValidAddress && (
          <TouchableOpacity
            testID="Contact-GetDirectionsButton"
            style={styles.actionButton}
            onPress={() =>
              Linking.openURL(
                `maps://?q=${encodeURIComponent(
                  address.address1 +
                    ', ' +
                    address.city +
                    ', ' +
                    address.state +
                    ' ' +
                    address.postcode,
                )}`,
              )
            }>
            <Text style={styles.actionButtonText}>{en.GetDirections}</Text>
          </TouchableOpacity>
        )}
        {email && (
          <TouchableOpacity
            testID="Contact-EmailButton"
            style={styles.actionButton}
            onPress={() => {
              try {
                const subject = 'Regarding your animal adoption listing';
                const body = `Hello,\n\nI am interested in adopting the following animal:\n\nName: ${selectedAnimal.name}\nID: ${selectedAnimal.id}\nURL: ${selectedAnimal.url}\n\nCould you please provide me with more information?`;
                Linking.openURL(
                  `mailto:${email}?subject=${encodeURIComponent(
                    subject,
                  )}&body=${encodeURIComponent(body)}`,
                );
              } catch (error) {
                console.error('Failed to open email:', error);
              }
            }}>
            <Text style={styles.actionButtonText}>{en.Email}</Text>
          </TouchableOpacity>
        )}
        {phone && (
          <TouchableOpacity
            testID="Contact-CallPhoneButton"
            style={styles.actionButton}
            onPress={() => Linking.openURL(`tel:${phone}`)}>
            <Text style={styles.actionButtonText}>{en.Call}</Text>
          </TouchableOpacity>
        )}
      </View>
    );
  };

  return (
    <View style={styles.container}>
      <Text style={styles.heading}>{en.ContactHeader}</Text>
      {renderContactInfo()}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f9f9f9',
    padding: 16,
    borderRadius: 8,
    marginVertical: 8,
  },
  heading: {
    fontSize: 18,
    fontWeight: 'bold',
    marginBottom: 8,
  },
  contactContainer: {
    marginVertical: 8,
  },
  actionButton: {
    backgroundColor: '#007AFF',
    padding: 10,
    borderRadius: 5,
    marginTop: 8,
    alignItems: 'center',
  },
  actionButtonText: {
    color: 'white',
    fontWeight: 'bold',
  },
});

export default Contact;
