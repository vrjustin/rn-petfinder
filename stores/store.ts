import {combineReducers, configureStore} from '@reduxjs/toolkit';
import petTypesReducer from '../reducers/petTypesReducer';
import petBreedsReducer from '../reducers/petBreedsReducer';
import animalsReducer from '../reducers/animalsReducer';
import searchParamsReducer from '../reducers/searchParamsReducer';

const createDebugger = require('redux-flipper').default;

const rootReducer = combineReducers({
  petTypes: petTypesReducer,
  petBreeds: petBreedsReducer,
  animals: animalsReducer,
  searchParameters: searchParamsReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

const store = configureStore({
  reducer: rootReducer,
  middleware: getDefaultMiddleware =>
    __DEV__
      ? getDefaultMiddleware({serializableCheck: false}).concat(
          createDebugger(),
        )
      : getDefaultMiddleware({serializableCheck: false}),
});

export default store;
