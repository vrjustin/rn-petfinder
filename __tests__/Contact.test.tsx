import React from 'react';
import {Provider} from 'react-redux';
import store from '../stores/store';
import renderer, {act} from 'react-test-renderer';
import Contact from '../Views/Contact';

let renderedContactTree: any;

describe('Contact', () => {
  it('renders correctly', async () => {
    await act(async () => {
      renderedContactTree = renderer.create(
        <Provider store={store}>
          <Contact
            route={{
              params: {
                selectedAnimal: {
                  id: 1,
                  organization_id: 'org1',
                  url: 'https://example.com',
                  age: 'Young',
                  gender: 'Female',
                  size: 'Medium',
                  coat: 'Short',
                  name: 'Fluffy',
                  description: 'A fluffy dog',
                  status: 'Available',
                  attributes: {
                    declawed: null,
                    house_trained: true,
                    shots_current: true,
                    spayed_neutered: false,
                    special_needs: false,
                  },
                  breeds: {
                    mixed: false,
                    primary: 'Poodle',
                    secondary: null,
                    unknown: false,
                  },
                  colors: {
                    primary: 'White',
                    secondary: null,
                    tertiary: null,
                  },
                  contact: {
                    address: {
                      address1: '123 Main St',
                      address2: null,
                      city: 'Anytown',
                      country: 'USA',
                      postcode: '12345',
                      state: 'NY',
                    },
                    email: 'email@example.com',
                    phone: '555-123-4567',
                  },
                  photos: [
                    {
                      full: 'https://example.com/full.jpg',
                      large: 'https://example.com/large.jpg',
                      medium: 'https://example.com/medium.jpg',
                      small: 'https://example.com/small.jpg',
                    },
                  ],
                  primary_photo_cropped: {
                    full: 'https://example.com/full.jpg',
                    large: 'https://example.com/large.jpg',
                    medium: 'https://example.com/medium.jpg',
                    small: 'https://example.com/small.jpg',
                  },
                  species: 'Dog',
                  tags: ['Friendly', 'Playful'],
                  isFavorite: false,
                },
              },
            }}
          />
        </Provider>,
      );
    });
    expect(renderedContactTree.toJSON()).toMatchSnapshot();
  });
});
