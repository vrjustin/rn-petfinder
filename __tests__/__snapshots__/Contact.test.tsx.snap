// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#f9f9f9",
      "borderRadius": 8,
      "flex": 1,
      "marginVertical": 8,
      "padding": 16,
    }
  }
>
  <Text
    style={
      {
        "fontSize": 18,
        "fontWeight": "bold",
        "marginBottom": 8,
      }
    }
  >
    Contact Information
  </Text>
  <View
    style={
      {
        "marginVertical": 8,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#007AFF",
          "borderRadius": 5,
          "marginTop": 8,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "fontWeight": "bold",
          }
        }
      >
        Get Directions
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#007AFF",
          "borderRadius": 5,
          "marginTop": 8,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "fontWeight": "bold",
          }
        }
      >
        Email
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#007AFF",
          "borderRadius": 5,
          "marginTop": 8,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "fontWeight": "bold",
          }
        }
      >
        Call
      </Text>
    </View>
  </View>
</View>
`;
